File: DSA/Linkedlist/Code/TraversalOfLinkedList.cs
// Traversal of Linked List in C#

using System;

class Node {
    public int data;
    public Node next;
    
    public Node(int data) {
        this.data = data;
        this.next = null;
    }
}

class LinkedListTraversal {
    Node head;
    
    void AddNode(int data) {
        Node newNode = new Node(data);
        if (head == null) {
            head = newNode;
            return;
        }
        Node temp = head;
        while (temp.next != null) temp = temp.next;
        temp.next = newNode;
    }
    
    void DisplayForward() {
        Console.Write("Forward: ");
        Node temp = head;
        while (temp != null) {
            Console.Write(temp.data + " -> ");
            temp = temp.next;
        }
        Console.WriteLine("NULL");
    }
    
    void DisplayRecursive(Node node) {
        if (node == null) {
            Console.WriteLine("NULL");
            return;
        }
        Console.Write(node.data + " -> ");
        DisplayRecursive(node.next);
    }
    
    int CountNodes(Node node) {
        if (node == null) return 0;
        return 1 + CountNodes(node.next);
    }
    
    Node GetNode(int pos) {
        if (pos < 1) return null;
        Node temp = head;
        for (int i = 1; i < pos && temp != null; i++)
            temp = temp.next;
        return temp;
    }
    
    static void Main() {
        LinkedListTraversal list = new LinkedListTraversal();
        
        Console.WriteLine("=== Traversal of Linked List ===\n");
        
        // Build list
        for (int i = 1; i <= 5; i++)
            list.AddNode(i * 10);
        
        // Test 1: Iterative forward traversal
        list.DisplayForward();
        
        // Test 2: Recursive traversal
        Console.Write("Recursive: ");
        list.DisplayRecursive(list.head);
        
        // Test 3: Count nodes
        Console.WriteLine("Total nodes: " + list.CountNodes(list.head));
        
        // Test 4: Get specific node
        Node node = list.GetNode(3);
        if (node != null)
            Console.WriteLine("Node at position 3: " + node.data);
        
        Console.WriteLine("\nComplexity Analysis:");
        Console.WriteLine("Forward traversal: O(n)");
        Console.WriteLine("Recursive traversal: O(n) time, O(n) space (call stack)");
        Console.WriteLine("Count nodes: O(n)");
        Console.WriteLine("Get at position: O(n)");
    }
}

File: DSA/Linkedlist/Code/TraversalOfLinkedList.cpp
#include <iostream>
using namespace std;

// Traversal of Linked List in C++
struct Node {
    int data;
    Node* next;
    Node(int data) : data(data), next(NULL) {}
};

class LinkedListTraversal {
private:
    Node* head;
    
public:
    LinkedListTraversal() : head(NULL) {}
    
    void addNode(int data) {
        Node* newNode = new Node(data);
        if (head == NULL) {
            head = newNode;
            return;
        }
        Node* temp = head;
        while (temp->next != NULL) temp = temp->next;
        temp->next = newNode;
    }
    
    void displayForward() {
        cout << "Forward: ";
        Node* temp = head;
        while (temp != NULL) {
            cout << temp->data << " -> ";
            temp = temp->next;
        }
        cout << "NULL\n";
    }
    
    void displayRecursive(Node* node = NULL) {
        static bool first = true;
        if (first) {
            cout << "Recursive: ";
            first = false;
            node = head;
        }
        if (node == NULL) {
            cout << "NULL\n";
            first = true;
            return;
        }
        cout << node->data << " -> ";
        displayRecursive(node->next);
    }
    
    int countNodes(Node* node = NULL) {
        if (node == NULL) node = head;
        if (node == NULL) return 0;
        return 1 + countNodes(node->next);
    }
    
    Node* getNode(int pos) {
        if (pos < 1) return NULL;
        Node* temp = head;
        for (int i = 1; i < pos && temp != NULL; i++)
            temp = temp->next;
        return temp;
    }
    
    ~LinkedListTraversal() {
        Node* temp;
        while (head != NULL) {
            temp = head;
            head = head->next;
            delete temp;
        }
    }
};

int main() {
    LinkedListTraversal list;
    
    cout << "=== Traversal of Linked List ===\n\n";
    
    // Build list
    for (int i = 1; i <= 5; i++)
        list.addNode(i * 10);
    
    // Test 1: Iterative forward traversal
    list.displayForward();
    
    // Test 2: Recursive traversal
    list.displayRecursive();
    
    // Test 3: Count nodes
    cout << "Total nodes: " << list.countNodes() << "\n";
    
    // Test 4: Get specific node
    Node* node = list.getNode(3);
    if (node != NULL)
        cout << "Node at position 3: " << node->data << "\n";
    
    cout << "\nComplexity Analysis:\n";
    cout << "Forward traversal: O(n)\n";
    cout << "Recursive traversal: O(n) time, O(n) space (call stack)\n";
    cout << "Count nodes: O(n)\n";
    cout << "Get at position: O(n)\n";
    
    return 0;
}
